src/go/k8s/controllers/redpanda/cluster_controller.go

```diff
@@ -129,62 +129,88 @@ func (r *ClusterReconciler) Reconcile(
 		}
 	}
 
+	err := r.reportStatus(ctx, &redpandaCluster, sts.LastObservedState, headlessSvc.HeadlessServiceFQDN(), nodeportSvc.Key())
+	if err != nil {
+		log.Error(err, "Unable to report status")
+	}
+
+	return ctrl.Result{}, err
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&redpandav1alpha1.Cluster{}).
+		Owns(&appsv1.StatefulSet{}).
+		Owns(&corev1.Service{}).
+		Complete(r)
+}
+
+func (r *ClusterReconciler) reportStatus(
+	ctx context.Context,
+	redpandaCluster *redpandav1alpha1.Cluster,
+	lastObservedSts *appsv1.StatefulSet,
+	internalFQDN string,
+	nodeportSvcName types.NamespacedName,
+) error {
 	var observedPods corev1.PodList
 
 	err := r.List(ctx, &observedPods, &client.ListOptions{
-		LabelSelector: labels.ForCluster(&redpandaCluster).AsClientSelector(),
+		LabelSelector: labels.ForCluster(redpandaCluster).AsClientSelector(),
 		Namespace:     redpandaCluster.Namespace,
 	})
 	if err != nil {
-		log.Error(err, "Unable to fetch PodList resource")
-
-		return ctrl.Result{}, err
+		return fmt.Errorf("unable to fetch PodList resource: %w", err)
 	}
 
 	observedNodesInternal := make([]string, 0, len(observedPods.Items))
 	// nolint:gocritic // the copies are necessary for further redpandacluster updates
 	for _, item := range observedPods.Items {
-		observedNodesInternal = append(observedNodesInternal, fmt.Sprintf("%s.%s", item.Name, headlessSvc.HeadlessServiceFQDN()))
+		observedNodesInternal = append(observedNodesInternal, fmt.Sprintf("%s.%s", item.Name, internalFQDN))
 	}
 
-	observedNodesExternal, err := r.createExternalNodesList(ctx, observedPods.Items, &redpandaCluster, nodeportSvc.Key())
+	observedNodesExternal, err := r.createExternalNodesList(ctx, observedPods.Items, redpandaCluster, nodeportSvcName)
 	if err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed to construct external node list: %w", err)
+		return fmt.Errorf("failed to construct external node list: %w", err)
 	}
 
-	if !reflect.DeepEqual(observedNodesInternal, redpandaCluster.Status.Nodes.Internal) ||
-		!reflect.DeepEqual(observedNodesExternal, redpandaCluster.Status.Nodes.External) {
-		redpandaCluster.Status.Nodes.Internal = observedNodesInternal
-		redpandaCluster.Status.Nodes.External = observedNodesExternal
-
-		if err := r.Status().Update(ctx, &redpandaCluster); err != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to update cluster status nodes: %w", err)
-		}
+	if lastObservedSts == nil {
+		return errNonexistentLastObservesState
 	}
 
-	if sts.LastObservedState == nil {
-		return ctrl.Result{}, errNonexistentLastObservesState
-	}
+	if statusShouldBeUpdated(redpandaCluster.Status, observedNodesInternal, observedNodesExternal, lastObservedSts.Status.ReadyReplicas) {
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
Why `DefaultRetry` and not `DefaultBackoff`? The error mentioned in cover latter was seen with how many instances of the operator running? 

I think that concurrent gorutines fights for updates to Cluster custom resource.

From the docs:
> // DefaultRetry is the recommended retry for a conflict where multiple clients
// are making changes to the same resource.
var DefaultRetry = wait.Backoff{
	Steps:    5,
	Duration: 10 * time.Millisecond,
	Factor:   1.0,
	Jitter:   0.1,
> }
>
> // DefaultBackoff is the recommended backoff for a conflict where a client
// may be attempting to make an unrelated modification to a resource under
// active management by one or more controllers.
var DefaultBackoff = wait.Backoff{
	Steps:    4,
	Duration: 10 * time.Millisecond,
	Factor:   5.0,
	Jitter:   0.1,
}

Reply:
I think there should not be any concurrent goroutines for two reasons:
- MaxConcurrentReconciles is set to 1 currently
- even if we had multiple, reconcile on one resource will never run multiple times in parallel, the workqueue makes sure that the one reconcile per key is run in parallel

These conflicts are created when:
- e.g. upgrade patches the sts and then we try to update status
- a user updates CRD between we patch the status (this will never happen in CI, will in real life though)

Reply:
re backoff: I don't think we need to really back off, as the documentation states it's if you have multiple controllers competing on resource, we don't have that

Reply:
I still don't get what causes this problem.

> e.g. upgrade patches the sts and then we try to update status

sts is different resource than Cluster custom resource. Maybe, because we set the Controller OwnerReference?